CMakeLists.txt

```diff
@@ -105,6 +105,28 @@ else()
   assert(0 "Unreachable code. Expected optimization level to have been chosen. See misc/Kconfig.")
 endif()
 
+# Dialects of C++, corresponding to the multiple published ISO standards.
+# Which standard it implements can be selected using the -std= command-line option.
+set_ifndef(DIALECT_STD_CPP98 "c++98")
+set_ifndef(DIALECT_STD_CPP11 "c++11")
+set_ifndef(DIALECT_STD_CPP14 "c++14")
+set_ifndef(DIALECT_STD_CPP17 "c++17")
+set_ifndef(DIALECT_STD_CPP2A "c++2a")
```

Comment:
Why not just have a CONFIG_STD_CPP_DIALECT with a string value (e.g. "c++11") instead of this mess?  It's no less general (you can always indirect on the value in the cmakelists.txt regardless of whether it's a symbol or string or integer or whatever) and MUCH less wordy.

Reply:
We want to retain the choice of options though.

A Kconfig choice is easier for users to input than a string (it is impossible for users to give invalid/unsupported input, e.g. is it -std=c++14, or c++14 ?) and it will be easier for us to make their input cross-toolchain portable.

If users can input arbitrary strings we would need to translate from a gcc name for the std to a non-gcc name. Translating from gcc to non-gcc is something that the TSC is probably soon going to disallow, in favour of intent-based abstractions.

Reply:
Note that the approach chosen here is the same as we use for the optimization flag -O<optimization>

Reply:
Do we?  It's not our list, it's GCC's.  We'd be doing nothing but transcribing it, and in practice transcribing the list badly by forgetting options.

If you want this to be error-checked, then we can run the compiler to validate it, no?

Reply:
We would be doing more than transcribe it, we would also be translating it from gcc to non-gcc when we add support for more toolchains.

And I believe that translating from gcc to non-gcc will be prohibited by the TSC soon.

Perhaps @mped-oticon can explain this better, he will be rewriting this later if everything goes according to plan.

Reply:
For my part I'm agree with @SebastianBoe.

I think that is easier for users and more safer, because user can't give invalid input.
It's also more easy for translating to another compiler.

C++ standards supported by GCC are limited (see [C++ Standards Support in GCC](https://www.gnu.org/software/gcc/projects/cxx-status.html)), so, for me, a list of choices is better.
I don't think I have forgotten options (C++03 is supported by c++98 option).